File: Python/Maximum Subarray.py
from sys import maxsize
from typing import List


class Solution:
    """
    Time:   O(n)
    Memory: O(1)
    """

    def maxSubArray(self, nums: List[int]) -> int:
        max_sum, curr_sum = -maxsize, 0

        for num in nums:
            curr_sum += num
            if max_sum < curr_sum:
                max_sum = curr_sum
            if curr_sum < 0:
                curr_sum = 0

        return max_sum

File: Java/Maximum Subarray.java
class Solution {

    /**
     * Time:   O(n)
     * Memory: O(1)
     */
    public static int maxSubArray(int[] nums) {
        int maxSum = nums[0], subarrayStart = nums[0];

        for (int i = 1; i < nums.length; ++i) {
            subarrayStart = Math.max(subarrayStart + nums[i], nums[i]);
            maxSum = Math.max(maxSum, subarrayStart);
        }

        return maxSum;
    }
}
